### PR TITLE
Epmlstrcmw 122

### PR DESCRIPTION
**Pull request by YuliyaSheyko**
Refine domain-level types and related logic with microtypes

### CODE REVIEW

File: portal/src/main/scala/cromwell/pipeline/datastorage/dao/entry/UserEntry.scala [hunk 1]
```diff
@@ -1,20 +1,26 @@
...
 
 trait UserEntry {
   this: Profile =>
 
+  import Implicits._
   import profile.api._
 
+//  implicit val SupplierIDColumnType = MappedColumnType.base[UserId, String](
+//    s => s.unwrap.toString, // map SupplierID to Int
+//    string => UserId(UUID.fromString(string)) // map Int to SupplierID
```
Comment (Dmitrii-Shevchenko) [inline portal/src/main/scala/cromwell/pipeline/datastorage/dao/entry/UserEntry.scala]: Remove comments

Comment (YuliyaSheyko) [reply portal/src/main/scala/cromwell/pipeline/datastorage/dao/entry/UserEntry.scala]: resolved

File: portal/src/main/scala/cromwell/pipeline/datastorage/dao/repository/UserRepository.scala [hunk 2]
```diff
@@ -2,8 +2,7 @@ package cromwell.pipeline.datastorage.dao.repository
 
 import cromwell.pipeline.database.PipelineDatabaseEngine
 import cromwell.pipeline.datastorage.dao.entry.UserEntry
-import cromwell.pipeline.datastorage.dto.User.UserEmail
-import cromwell.pipeline.datastorage.dto.{ User, UserId }
+import cromwell.pipeline.datastorage.dto.{ User, UserEmail, UUID => UserId }
```
Comment (ilyaonishenko) [inline portal/src/main/scala/cromwell/pipeline/datastorage/dao/repository/UserRepository.scala]: We should use UUID here, not UserId

File: portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala [hunk 3]
```diff
@@ -1,36 +1,76 @@
...
+  type Type = String
+  type Wrapper = Name
+  type Error = String
+  implicit lazy val nameFormat: Format[Name] = Json.valueFormat[Name]
+  override protected def create(value: String): Name = new Name(value)
+  override protected def validate(value: String): ValidationResult[String] =
+    Validated.cond(
+      value.matches("^[a-zA-Z]+$"),
+      value,
+      NonEmptyChain.one("Name cannot contain spaces, numbers or special characters.sbt.")
```
Comment (ilyaonishenko) [inline portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala]: `.sbt.`?

Comment (ilyaonishenko) [reply portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala]: Redefine message. Name should contain only letters (but make it more beautiful, please)

File: portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala [hunk 4]
```diff
@@ -1,36 +1,76 @@
...
+    )
+}
+
+final class UUID(override val unwrap: String) extends AnyVal with Wrapped[String]
```
Comment (ilyaonishenko) [inline portal/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala]: private constructor

File: portal/src/main/scala/cromwell/pipeline/datastorage/dto/auth/SignUpRequest.scala [hunk 5]
```diff
@@ -1,9 +1,34 @@
...
+}
+
+object Password extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = Password
+  type Error = String
+  implicit lazy val passwordFormat: Format[Password] = Json.valueFormat[Password]
+  override protected def create(value: String): Password = new Password(value)
+  override protected def validate(value: String): ValidationResult[String] = Validated.cond(
+    value.matches("(?=^.{10,}$)((?=.*\\d)|(?=.*\\W+))(?![.\\n])(?=.*[A-Z])(?=.*[a-z]).*$"),
```
Comment (ilyaonishenko) [inline portal/src/main/scala/cromwell/pipeline/datastorage/dto/auth/SignUpRequest.scala]: [NAI] Improve password check

File: portal/src/main/scala/cromwell/pipeline/datastorage/utils/ColumnTypes.scala [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package cromwell.pipeline.datastorage.utils
+
+import cromwell.pipeline.datastorage.dto.{ Name, UUID, UserEmail }
+import cromwell.pipeline.datastorage.Profile
+
+trait ColumnTypes {
```
Comment (ilyaonishenko) [inline portal/src/main/scala/cromwell/pipeline/datastorage/utils/ColumnTypes.scala]: Remove please

File: portal/src/main/scala/cromwell/pipeline/utils/validator/Wrapped.scala [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    type Type
+    type Error
+    type Wrapper <: Wrapped[Type]
+    type ValidationResult[A] = Validated[NonEmptyChain[Error], A]
+    implicit def wrappedOrdering(implicit ord: Ordering[Type]): Ordering[Wrapper] = Ordering.by(_.unwrap)
+    implicit def unwrap(wrapped: Wrapper): Type = wrapped.unwrap
+    protected def create(value: Type): Wrapper
+    protected def validate(value: Type): ValidationResult[Type]
+    def from(value: Type): ValidationResult[Wrapper] =
+      validate(value) match {
```
Comment (ilyaonishenko) [inline portal/src/main/scala/cromwell/pipeline/utils/validator/Wrapped.scala]: `validate(value).map(create)`

File: portal/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala [hunk 8]
```diff
@@ -62,23 +71,32 @@ class AuthControllerTest extends WordSpec with Matchers with MockFactory with Sc
         }
       }
 
-      "return BadRequest with fields validation errors" in {
-        val signUpRequest = SignUpRequest("email", "password", "First-name", "Last-name")
+      "return BadRequest with fields validation errors" ignore {
```
Comment (ilyaonishenko) [inline portal/src/test/scala/cromwell/pipeline/controller/AuthControllerTest.scala]: Add todo with reference to Jira ticket

File: portal/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala [hunk 9]
```diff
@@ -58,7 +59,7 @@ class ProjectServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
         val project =
           Project(
             projectId = projectId,
-            ownerId = UserId("userId"),
+            ownerId = UUID("123e4567-e89b-12d3-a456-426655440000"),
```
Comment (ilyaonishenko) [inline portal/src/test/scala/cromwell/pipeline/service/ProjectServiceTest.scala]: Please use `UUID.random`

File: portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala [hunk 10]
```diff
@@ -50,12 +57,28 @@ class UserServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
...
-        val request = UserUpdateRequest("updatedEmail", "updatedFirstName", "updatedLastName")
+        val userId = UUID("123e4567-e89b-12d3-a456-426655440000")
+        val user = User(
+          userId,
+          UserEmail("someDomain@cromwell.com"),
+          "hash",
+          "salt",
+          Name("name"),
+          Name("lastName"),
+          active = false
```
Comment (ilyaonishenko) [inline portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Remove this line please

File: portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala [hunk 11]
```diff
@@ -66,14 +89,21 @@ class UserServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
...
         val user =
-          User(UserId(id), "email@cromwell.com", calculatePasswordHash("password", salt), salt, "name", "lastName")
+          User(
+            id,
+            UserEmail("email@cromwell.com"),
+            calculatePasswordHash("password", salt),
+            salt,
+            Name("name"),
+            Name("lastName")
+          )
```
Comment (ilyaonishenko) [inline portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Use named parameters

File: portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala [hunk 12]
```diff
@@ -66,14 +89,21 @@ class UserServiceTest extends AsyncWordSpec with Matchers with MockitoSugar {
 
     "updatePassword" should {
       "returns success if database handles query" in {
-        val id = "123"
+        val id = UUID("123e4567-e89b-12d3-a456-426655440000")
```
Comment (ilyaonishenko) [inline portal/src/test/scala/cromwell/pipeline/service/UserServiceTest.scala]: Use `UUID.random`

**Comment by Dmitrii-Shevchenko** [comment]
https://github.com/EpamLifeSciencesTeam/cmwl_pipeline/pull/44

**Pull request closed without merging**